cityofaustin/atd-vz-data
# Replace lambda function to make sure austin_full_purpose (AFP) is flagged correctly

## Associated issues
Closes https://github.com/cityofaustin/atd-data-tech/issues/10126

First of all, I apologize for the length of this!

This PR addresses pulling the functionality for updating the `austin_full_purpose` flag out of AWS lambda which looks like it has not been servicing its purpose lately (I tested updating crash coordinates to be outside of Austin/vice versa through the VZE UI and the flag wasn't updating). We instead want the flag to be handled directly in the DB which will be easier to maintain. The route I ended up taking for this was creating a stored generated column called `in_austin_full_purpose` that calls a function `is_in_austin_full_purpose` whenever a record is added to the `atd_txdot_crashes` table or when a record is updated. The function returns true/false for whether the crash is in AFP and then that value is stored in the column.

_Cool thing about stored generated columns from the postgres docs:
"A stored generated column is computed when it is written (inserted or updated) and occupies storage as if it were a normal column"_

The determination for whether a crash is in AFP is based off this message that Xavier sent us in the past:
["The austin_full_purpose flag should only be "Y" when it is in jurisdiction = 5 OR when the crash does not have a latitude/longitude and the city_id = 22"](https://austininnovation.slack.com/archives/CM9MK950S/p1678398208097029)

The AFP flag being correct will ensure that crashes are showing up in the Crashes page correctly and the same goes for the Fatalities page since the view that powers that page uses the AFP flag to determine what fatalities make it into the view. This will also help with making sure fatality records have an engineering area, because previously we had some fatalities that were showing up in the Fatalities page that didn't have engineering areas because they werent actually in Austin. With that being said, there are still about 7 fatalities where the engineering area is NULL because it seems like the engineering areas dont perfectlyyy line up with the AFP jurisdiction, and I'm talking like meters off, so its definitely an edge case thing and I'm not sure if thats a problem or how to handle it.

Heres an example of that:

The crash id 13511720 has a fatality and is in the AFP jurisdiction but it just barely falls outside of what would be its fatality engineering area, so the record is showing up in the Fatality view as it should, but it has a null engineering area.

AFP Jurisdiction layer (includes the right lane of Dessau rd):
![Screenshot from 2023-06-27 16-07-21](https://github.com/cityofaustin/atd-vz-data/assets/67130963/be468cb7-b26a-4ee1-a4ea-e92a461a4328)

Engineering area layer (does not include the right lane of Dessau rd):
![Screenshot from 2023-06-27 16-07-38](https://github.com/cityofaustin/atd-vz-data/assets/67130963/842ae216-892e-41e1-82f6-bca0445e7a71)

Where the crash is located (on the right lane of Dessau rd):
![Screenshot from 2023-06-28 17-44-55](https://github.com/cityofaustin/atd-vz-data/assets/67130963/e59fbb06-1534-4304-9db3-cd042a39cabc)



## Testing
**URL to test:** 
https://deploy-preview-1247--atd-vze-staging.netlify.app/#/fatalities

**Steps to test:**
1. Go to the Fatalities page, observe that its rendering without errors and the engineering areas are there for mostly all the fatalities.
2. Go to the Crashes page, toggle on and off the AFP toggle.
3. Mostly though look at the code and let me know what you think of this solution!

Steps moving forward:

- [ ] Retire the `austin_full_purpose` column?
- [x] Need to address the other part of what the lambda function did, which was updating the `city_id`. That should be its own issue. And this probably shouldn't be merged until that is done and we can shut down the lambda function all at once?
- [ ] Run all these commands in prod before merge.
- [ ] Make sure all views and the socrata etl is using the new `in_austin_full_purpose` column
- [ ] Delete the `notify_hasura_jurisdiction_from_latlon_INSERT` - AFTER INSERT
`notify_hasura_jurisdiction_from_latlon_UPDATE` - AFTER UPDATE triggers
- [ ] Retire the `jurisdiction_from_latlon` hasura event
- [ ] Shut down the lambda functions

---
#### Ship list
